components/frontend/src/App.js

```diff
@@ -152,24 +145,47 @@ class App extends Component {
         login("", "")
             .then(function (json) {
                 if (json.ok) {
-                    self.set_user(json.email, json.email, json.session_expiration_datetime);
+                    self.setUserSession(json.email, json.email, new Date(json.session_expiration_datetime));
                     return true;
                 }
             });
         return false;
     }
 
-    set_user(username, email, session_expiration_datetime) {
-        const email_address = email && email.indexOf("@") > -1 ? email : null;
-        this.setState({ user: username, email: email_address });
-        if (username === null) {
+    initUserSession() {
+        // Check if there is a session expiration datetime in the local storage. If so, restore the session as long as
+        // it has not expired. Otherwise, nothing needs to be done.
+        const sessionExpirationDateTimeISOString = localStorage.getItem("session_expiration_datetime")
+        if (sessionExpirationDateTimeISOString) {
+            const sessionExpirationDateTime = new Date(sessionExpirationDateTimeISOString)
+            if (sessionExpirationDateTime < new Date()) {
+                // The session expired while the user was away. Reset it and notify the user of the expired session.
+                this.setUserSession()
+                show_message("warning", "Your session expired", "Please log in to renew your session");
+            } else {
+                // Session is still active, restore it from local storage.
+                this.setUserSession(localStorage.getItem("user"), localStorage.getItem("email"), sessionExpirationDateTime)
+            }
+        }
+    }
+
+    setUserSession(username, email, sessionExpirationDateTime) {
+        if (username) {
+            const emailAddress = email && email.indexOf("@") > -1 ? email : null;
+            this.setState({ user: username, email: emailAddress });
+            localStorage.setItem("user", username);
+            localStorage.setItem("email", emailAddress);
+            localStorage.setItem("session_expiration_datetime", sessionExpirationDateTime.toISOString());
+            this.sessionExpirationTimeoutId = setTimeout(() => {
+                this.setUserSession(null);
+                show_message("warning", "Your session expired", "Please log in to renew your session");
+            }, sessionExpirationDateTime - new Date())
+        } else {
+            this.setState({ user: null, email: null });
```

Comment:
Why is `setState` sometimes followed by reload but sometimes not?

Reply:
Good question. I'll see if I can add comments.